### PR TITLE
Add Damage net event

### PR DESCRIPTION
**Pull request by Dune-jr**
So this is a proposal to include a `Damage` net event.
This event contains the following:

- `m_ClientID`: ID of damaged tee
- `m_Amount`: amount of damage taken, from 1 to 9
- `m_Angle`: angle from which damage was taken
- `m_Armored`: type of damage (armor damage or health damage) 

This would allow for some later potential graphical improvements that would need to take into account the affected tee, and the angle from which the damage came.

I also added the boolean `m_Armored` flag to `DamageInd`, in case we want to keep server-side damage indicators for health-only.

This is related to #1715, which is only a foretaste of what could be done ;-)
Adding the pre-feature of #1715 is completely optional, I have it done and ready though.

Note: Actually, it should also be possible to completely remove the `DamageInd` packet and generate the damage indicators from the client. I'm not sure if this has any unwanted side-effects. 
PS: the `m_Angle` flag was tested with all weapons/particles.

### CODE REVIEW

**Comment by oy** [comment]
What's the client_id for? And what about the Amount?
Should add as less entries as possible - creates all traffic.

Instead of armor I'd add Type so self damage could be indicated too #1411.

**Comment by Dune-jr** [comment]
> What's the client_id for? And what about the Amount?

Right now, it is not possible to know which tee was damaged, afaik. So without the clientID, you cannot do anything relative to the position of the tee.

> Should add as less entries as possible - creates all traffic.

Well then I'll remove the `NetBool("m_Armored"),` from DamageInd, it could be worked around. I don't see how to scrap away more entries though :/

Handling the damage indicators from the client would be maybe cleaner, but I don't think it would save traffic.

> Instead of armor I'd add Type so self damage could be indicated too #1411.

Ah, good point. It would be just as compact to add a `m_Self` boolean, right?

**Comment by oy** [comment]
> 
> 
> > What's the client_id for? And what about the Amount?
> 
> Right now, it is not possible to know which tee was damaged, afaik. So without the clientID, you cannot do anything relative to the position of the tee.

Ah ok. 

> 
> > Should add as less entries as possible - creates all traffic.
> 
> Well then I'll remove the `NetBool("m_Armored"),` from DamageInd, it could be worked around. I don't see how to scrap away more entries though :/
> 
> Handling the damage indicators from the client would be maybe cleaner, but I don't think it would save traffic.

Well, if they have a purpose it's ok to add them. But should prevent redundant information. Dunno if Amount is relevant?
> 
> > Instead of armor I'd add Type so self damage could be indicated too #1411.
> 
> Ah, good point. It would be just as compact to add a `m_Self` boolean, right?
Would go with int m_Type  and then would set a DamageEnum in network.py and NORMAL, ARMOR, SELF

**Comment by Dune-jr** [comment]
> > Well then I'll remove the `NetBool("m_Armored"),` from DamageInd, it could be worked around. I don't see how to scrap away more entries though :/
> > Handling the damage indicators from the client would be maybe cleaner, but I don't think it would save traffic.
> 
> Well, if they have a purpose it's ok to add them. But should prevent redundant information. Dunno if Amount is relevant?

Well, we could remove `m_Amount` and send one packet per damage like it is done with the damage indicators. That sounds like *more* traffic though, right?

> > > Instead of armor I'd add Type so self damage could be indicated too #1411.
> > 
> > Ah, good point. It would be just as compact to add a `m_Self` boolean, right?
>
> Would go with int m_Type  and then would set a DamageEnum in network.py and NORMAL, ARMOR, SELF

I don't know too much about the internals, but am I not correct in saying that this doesn't save any space compared to two booleans? It has the disadvantage that self normal and self armor are not distinguished (although that may not be necessary).

**Comment by Dune-jr** [comment]
Fixed. I added `SELF_ARMOR` as well in case, not sure if we will need that.

**Comment by fokkonaut** [comment]
maybe rename to m_Armored to m_Type and make two new variables named DAMAGE_ARMOR, DAMAGE_HEALTH

**Comment by Dune-jr** [comment]
> maybe rename to m_Armored to m_Type and make some new variables named for example DAMAGE_ARMOR, DAMAGE_HEALTH, DAMAGE_BOTH

Look at the commit, please. There is already no more m_Armored, only m_Type.

File: src/game/server/entities/character.cpp [hunk 1]
```diff
@@ -729,6 +718,19 @@ bool CCharacter::TakeDamage(vec2 Force, int Dmg, int From, int Weapon)
 		m_Health -= Dmg;
 	}
 
+	// create healthmod indicator
+	if(Server()->Tick() < m_DamageTakenTick+25)
+	{
+		// make sure that the damage indicators doesn't group together
+		GameServer()->CreateDamageInd(m_Pos, m_DamageTaken*0.25f, Dmg);
```
Comment (oy) [inline src/game/server/entities/character.cpp]: Creates and sends 2 events right now. We should either use the new event or just send the old one.

Comment (Dune-jr) [reply src/game/server/entities/character.cpp]: Right now the server decides the position, angle, etc. of the stars, so we need this kind of event, unless the whole damageind code is ported to the client.

If we only send the old one, this means any graphical effect designed in 0.7.2 will not work with 0.7.1 servers. I'm fine with that if you think that is better than sending currently unused packets, you tell me (code-wise this means removing the CreateDamage server calls for now).

Comment (oy) [reply src/game/server/entities/character.cpp]: Might be best to drop the old CreateDamageInd and just use the new event.

Comment (Dune-jr) [reply src/game/server/entities/character.cpp]: Did just that.

File: src/game/server/gamecontext.cpp [hunk 2]
```diff
@@ -93,46 +93,19 @@ class CCharacter *CGameContext::GetPlayerChar(int ClientID)
...
-				else
-					pEvent->m_Type = ArmorType ? DAMAGE_SELF_ARMOR : DAMAGE_SELF;
-			}
-		}
+		pEvent->m_X = (int)Pos.x;
+		pEvent->m_Y = (int)Pos.y;
+		pEvent->m_ClientID = Id;
+		pEvent->m_Angle = (int)(f*256.0f);
+		pEvent->m_HealthAmount = HealthAmount;
+		pEvent->m_ArmorAmount = HealthAmount;
```
Comment (oy) [inline src/game/server/gamecontext.cpp]: Typo?

Comment (Dune-jr) [reply src/game/server/gamecontext.cpp]: Ugh. Fixed.

**Pull request merged**